src/include/sof/measure.h

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2019, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+ */
+
+#ifndef __MEASURE_H__
+#define __MEASURE_H__
+
+#include <stdint.h>
+#include <sof/alloc.h>
+#include <sof/clk.h>
+#include <sof/platform.h>
+#include <sof/drivers/timer.h>
+
+#if CONFIG_PERF_MEASURE
+
+/**
+ * Trace appendix containing default information
+ */
+#define PERF_REPORT_FMT " - avg perf: %u us, %u cycles"
+
+/**
+ * Statistics types
+ */
+enum measure_stats {
+	MEASURE_STAT_MIN = 0,
+	MEASURE_STAT_AVG,
+	MEASURE_STAT_MAX,
+
+	MEASURE_NUM_STATS
+};
+
+/**
+ * Measurement sample
+ */
+struct measure_sample {
+	uint64_t time_delta;
+	uint32_t ccount_delta;
+};
+
+/**
+ * Measurement context
+ */
+struct measure_context {
+	uint32_t steps;
+	uint32_t avg_shift;
+	uint32_t cur_step;
+	uint64_t begin_time;
+	uint64_t begin_ccount;
+	uint64_t last_report;
+	struct measure_sample *samples;
+	struct measure_sample stats[MEASURE_NUM_STATS];
+};
+
+/**
```

Comment:
I would define the enum and structs outside of #if CONFIG_PERF_MEASURE. Btw should it be #ifdef?

Reply:
@dbaluta If you ask about `#if CONFIG_*` vs `#if defined(CONFIG_*)`, then I recommended option without defined, because `#if defined()` sucks in cases when your symbol is for example `#define CONFIG_SOMETHING 0`, it would work when probably you don't want that. However I saw that in kernel they use both versions.

Reply:
@dbaluta tho, it would be the best to have IS_ENABLED() like in kernel, but I don't know how we gonna support modules, so we use `#if symbol` for now.

Reply:
@jajanusz i see your point, lets use #if symbol for now.